### DISCUSSION TITLE
How can I query encrypted values?

### DISCUSSION THREAD

**Discussion by ojtramp** [sequelize/sequelize#16372] [Help & Questions] [answered]
I would like to encrypt certain information (i.e. emails) within my database, but can't hash them as I also need to query them and use the raw value again in the future, sometimes without user input. Initially I used getters and setters but have since been advised to use hooks, I think that my issue with the hooks is that the encrypted value is different each time (due to the use of a random iv in AES-GCM) so I can't simply encrypt my search value and compare it to the encrypted values in the database.

**How can I keep my data encrypted at rest within the database, but also query it using `.findAll({where:{ email: 'example@gmail.com'}});`?**

Thanks everyone for the help

**Using getters and setters**

```
const aes256gcm = require('../services/aes256gcm');

...
email: {
  type: DataTypes.STRING,
  get() {
    return aes256gcm.aes256gcm(process.env.ENCRYPTION_KEY).decrypt(storedValue);
  },
  set(value) {
    this.setDataValue('email', aes256gcm.aes256gcm(process.env.ENCRYPTION_KEY).encrypt(value));
  }
},
...
```

**Using hooks**

    hooks : {
      afterFind : (loginMethod, options) => {
        loginMethod.email = aes256gcm.aes256gcm(process.env.ENCRYPTION_KEY).decrypt(loginMethod.email);
      },
      beforeFind : (loginMethod, options) => {
        loginMethod.email = aes256gcm.aes256gcm(process.env.ENCRYPTION_KEY).encrypt(loginMethod.email);
      },
      beforeCreate : (loginMethod, options) => {
        loginMethod.email = aes256gcm.aes256gcm(process.env.ENCRYPTION_KEY).encrypt(loginMethod.email);
      },
      beforeUpdate : (loginMethod, options) => {
        loginMethod.email = aes256gcm.aes256gcm(process.env.ENCRYPTION_KEY).encrypt(loginMethod.email);
      },
      beforeBulkCreate : (loginMethod, options) => {
        loginMethod.email = aes256gcm.aes256gcm(process.env.ENCRYPTION_KEY).encrypt(loginMethod.email);
      },
      beforeBulkUpdate : (loginMethod, options) => {
        loginMethod.email = aes256gcm.aes256gcm(process.env.ENCRYPTION_KEY).encrypt(loginMethod.email);
      },

    },

**Comment by ephys** [comment, marked as answer]
Hooks are a concept of the JavaScript layer, but your query needs decrypting on the SQL layer. You need to decrypt on the SQL layer. This depends on which database you're using.

You'll need to learn about writing raw SQL for this: 
- v7: https://sequelize.org/docs/v7/querying/raw-queries/
- v6: https://sequelize.org/docs/v6/core-concepts/raw-queries/

You'll also need to look up your database's documentation for AES

```ts
const email = 'example@gmail.com';

// Example Solution for Sequelize 7

User.findAll({
  where: sql`AES_DECRYPT(${sql.attribute('email')}, ${process.env.ENCRYPTION_KEY}) = ${email}`,
});

// Example Solution for Sequelize 6

User.findAll({
  where: where(fn('AES_DECRYPT', col('email'), process.env.ENCRYPTION_KEY), '=', email),
});
```
